### PR TITLE
8282431: AArch64: Add optimized rules for masked vector multiply-add/sub for SVE

### PR DESCRIPTION
**Pull request by XiaohongGong**
We have the optimized match rules for vector `"fmls,fnmla,fnmls,mla,mls"` for ARM SVE. Similarly we can also add the rules for the relative masked operations to generate the optimized predicated instructions.

With this patch the following masked vector multiply-add for a byte vector:
```
  mul z18.b, p7/m, z18.b, z17.b
  add z17.b, p0/m, z17.b, z18.b
```
could be optimized to: 
```
   mla z19.b, p0/m, z18.b, z17.b
```
 And so does the multiply-sub operations. Also the following masked fused multiply-substract for a float vector:
```
  fneg    z18.s, p7/m, z18.s
  fmad    z17.s, p0/m, z18.s, z16.s
```
could be optimized to: 
```
  fmsb z17.s, p0/m, z18.s, z16.s"
```
 And the same to the relative negated fused operations.

This patch also fixes the potential issues for the usage of `NegVF/D` in match rules. The explicit check of non-predicated vector for `NegVF/D` must be added to the match rule predicate if the `NegVF/D` is assumed to be non-masked. Otherwise, the jvm might crash if a masked` NegVF/D` with two operands is matched into a rule which assumes the `NegVF/D` in subtree has one operand.

Also add the jtreg tests for all the touched vector operations.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282431](https://bugs.openjdk.java.net/browse/JDK-8282431): AArch64: Add optimized rules for masked vector multiply-add/sub for SVE

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer) ⚠️ Review applies to 20f1e10650f5eefe6c0e48e36d78173297d28fd2
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7737/head:pull/7737` \
`$ git checkout pull/7737`

Update a local copy of the PR: \
`$ git checkout pull/7737` \
`$ git pull https://git.openjdk.java.net/jdk pull/7737/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7737`

View PR using the GUI difftool: \
`$ git pr show -t 7737`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7737.diff">https://git.openjdk.java.net/jdk/pull/7737.diff</a>

</details>

### CODE REVIEW

**Comment by XiaohongGong** [comment]
> Did you consider writing a test using the IR test framework?

Thanks for looking at this patch @PaulSandoz ! Yeah, I considered using the IR test framework. The main reason that I didn't use it finally is that this patch only changes the generated codes by adding match rules with out no IR changes.  And I misunderstood that the ir framework can only check the C2 IR nodes. Seems it can also check the assembly codes. I will change the tests to use it. Thanks!

**Comment by PaulSandoz** [comment]
> And I misunderstood that the ir framework can only check the C2 IR nodes. Seems it can also check the assembly codes.

To be honest I was uncertain if the ir framework could do this. Worth investigating.

**Comment by XiaohongGong** [comment]
> > And I misunderstood that the ir framework can only check the C2 IR nodes. Seems it can also check the assembly codes.
> 
> To be honest I was uncertain if the ir framework could do this. Worth investigating.

Yeah, since it checks the output of `-XX:+PrintOptoAssembly`, it can got the information from the "format" in a rule. I'v updated the test to check the SVE assembler codes. Thanks!

**Comment by XiaohongGong** [comment]
Hi, could anyone please help to take a look at this PR? Thanks so much!

**Review by nsjian — APPROVED** [review]

**Comment by XiaohongGong** [comment]
Hi, could anyone please help to take a look at this PR? Thanks so much!

File: test/hotspot/jtreg/compiler/vectorapi/VectorFusedMultiplyAddSubTest.java [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+           if (m[i % D_SPECIES.length()]) {
+               Asserts.assertEquals(f.apply(a[i], b[i], c[i]), r[i]);
+           } else {
+               Asserts.assertEquals(a[i], r[i]);
+           }
+       }
+    }
+
+    @Test
+    @IR(failOn = { "sve_add", "sve_mul" })
```
Comment (nick-arm) [inline test/hotspot/jtreg/compiler/vectorapi/VectorFusedMultiplyAddSubTest.java:226]: Do you also want something like `@IR(counts = { "sve_mla", 1 })` here to make sure the fused instruction is generated, in addition to not generating the two un-fused instructions?

Comment (XiaohongGong) [reply test/hotspot/jtreg/compiler/vectorapi/VectorFusedMultiplyAddSubTest.java:226]: Yeah, I think `@IR(counts = { "sve_mla", ">=1" })` is better. I will update the tests. Thanks for the review!

**Comment by XiaohongGong** [comment]
Hi @nick-arm , I'v updated the tests, could you please take a look at it again? Thanks!

**Review by nick-arm — APPROVED** [review]

**Comment by XiaohongGong** [comment]
Thanks for the review @nsjian @nick-arm @PaulSandoz !

**Pull request closed without merging**